File: lib/browser.js
var process = require('./process');

var cmds = [
  'get',
  'setHtml',
  'run',
  'runAsync',
  'title',
  'url',
  'html',
  'save',
  'screenshot'
];

exports.createBrowser = function(opts) {
  return new Browser(opts);
};

var Browser = function(opts) {
  this.id = 0;
  this.callbacks = {};
  var connection = this.connection = process.createProcess(opts);
  var self = this;
  connection.on('message', function(rv) {
    if(rv.error) {
      rv.error = new Error(rv.error);
    }
    if(self.callbacks[rv.id] && typeof(self.callbacks[rv.id]) == "function") {
      self.callbacks[rv.id](rv.error, rv.result);
    }
    delete self.callbacks[rv.id];
    if((self.queue && Object.keys(self.callbacks).length == 0) || rv.error) {
      self.exit();    
      if(self.endCallback) { 
        self.endCallback(rv.error, rv.result);
      }
    }
  });
  connection.on('error', function(err) {
    self.exit();
    if(self.endCallback) {
      self.endCallback(err);
    } else {
      throw err;
    }
  });
};

Browser.prototype.__defineGetter__('chain', function() {
  this.queue = [];
  return this;
});

Browser.prototype.command = function(cmd, args) {
  var id = ++this.id;
  this.callbacks[id] = (typeof(args[args.length-1]) == "function") ? args.pop() : true;
  var opts = { 'method' : cmd, 'params' : args, 'id' : id };
  this.queue ? this.queue.push(opts) : this.connection.send(opts);
};

Browser.prototype.end = function(cb) {
  this.endCallback = cb;
  this.connection.send(this.queue);
};

Browser.prototype.exit = function() {
  this.connection.close();
};

cmds.map(function(cmd) {
  Browser.prototype[cmd] = function() {
    var args = Array.prototype.slice.call(arguments);
    if(['run', 'runAsync'].indexOf(cmd) !== -1 && typeof(args[0]) == 'function') {
      args[0] = '('+args[0].toString()+')()';
    }
    this.command(cmd, args);
    return this;
  };
});

File: src/browser.cpp
#include <QApplication>
#include <QDebug>
#include <QObject>
#include <QVariant>
#include <QString>
#include <iostream>

#include "browser.h"
#include "json.h"

Browser::Browser() {
  QObject::connect(&reader, SIGNAL(commandReceived(QVariant)), this, SLOT(onCommandReceived(QVariant)));
  QObject::connect(&page, SIGNAL(loadFinished(bool)), this, SLOT(onLoadFinished(bool)));
  QObject::connect(&page.jsCallbackObject, SIGNAL(asyncScriptCallback(QVariant)), this, SLOT(onAsyncScriptCallback(QVariant)));  
  isRunning = false;
}

void Browser::next() {
  if(queue.isEmpty() || isRunning) {
    return;
  }
  runCommand(queue.dequeue());
}

void Browser::runCommand(QVariant cmd) {
  QMap<QString, QVariant> command = cmd.toMap();
  QVariantList params = command.value("params").toList();
  
  method = command.value("method").toString();
  id = command.value("id");
  isRunning = true;
  
  if(method == "run") {
    sendResponse(page.run(params.at(0).toString()));
  } else if(method == "save") {
    sendResponse(page.save(params.at(0).toString()));
  } else if(method == "html") {
    sendResponse(page.html());   
  } else if(method == "runAsync") {
    page.run(params.at(0).toString());
  } else if(method == "get") {
    page.get(params.at(0).toString());
  } else if(method == "setHtml") {
    page.setHtml(params.at(0).toString());
  } else {
    sendError(QString("Unknown Command"));
  }
}

void Browser::sendResponse(const QVariant &response) {
  QMap<QString, QVariant> json;
  json["result"] = response;
  json["id"] = id;
  std::cout << Json::stringify(json).data() << std::endl << std::flush;
  isRunning = false;
  next();
}

void Browser::sendError(const QString &response) {
  QMap<QString, QVariant> json;
  json["error"] = response;
  json["id"] = id;
  std::cout << Json::stringify(json).data() << std::endl << std::flush;
  QApplication::exit(1);
}

void Browser::onCommandReceived(QVariant cmd) {
  queue.enqueue(cmd);
  next();
}

void Browser::onLoadFinished(bool ok) {
  if(method == "get" || method == "setHtml") {
    sendResponse(ok);
  }
}

void Browser::onAsyncScriptCallback(QVariant result) {
  if(method == "runAsync") {
    sendResponse(result);
  }
}
